crates/ra_hir_ty/src/display.rs

```diff
@@ -81,36 +133,120 @@ impl<'a, 'b> HirFormatter<'a, 'b> {
     }
 }
 
-pub struct HirDisplayWrapper<'a, T>(&'a dyn HirDatabase, &'a T, Option<usize>, bool);
+pub trait RenderTarget: Sized + Copy {
```

Comment:
hmm I think this will be a bit too limited in what behavior it needs to actually customize (e.g. we'll want to render closure types differently), and I'm not sure we really need a trait anyway. We could just start with an `enum { RenderSourceCode { module_id }, RenderDiagnostics }`?

Reply:
The reason I went with traits is that I wanted to keep the return type of the existing `HirDisplay` calls the same while adding new errors to the new call. I don't think that's possible if the two variants are in the same enum?

Reply:
Yeah, I think it'd be fine to use the same error type for both.